File: 014/longest-chain.clj
#!/usr/bin/env clojure
(defn collatz [start]
  (defn next-collatz [n]
    (if (even? n)
      (/ n 2)
      (+ (* 3 n) 1)))
  (def memo-collatz
    (memoize next-collatz))
  (defn not-one? [n]
    (not (= n 1)))
  (concat (take-while not-one? (iterate next-collatz start)) [1]))

(defn collatz-seqs [limit]
  (map collatz (range 1 limit)))

(println (apply max-key count (collatz-seqs 100000)))

File: 014/longest-chain.rb
#!/usr/bin/env ruby

max_l = 0
max_i = 0
500001.step(1000000, 2).each do |i|
  l = 0
  j = i
  while j != 1 do
    if j.even?
      j /= 2
    else
      j = 3 * j + 1
    end
    l += 1
  end
  if l > max_l
    max_l = l
    max_i = i
  end
end

puts max_i
